tests/test_pep518.py

```diff
@@ -0,0 +1,21 @@
+import pytest
+
+from .conftest import cd_package
+
+
+@pytest.mark.parametrize(
+    ('package'),
+    [
+        'scipy-like',
+    ]
+)
```

Comment:
I'm not so familiar with how pytest parametrized tests work, but what is the point of having a test parameter that is always the same? I'm also not sure I grok what this test is doing. It installs the `build` package in a virtualenv, but `build` has been added to the test dependencies. Why can't we just use `subprocess.run()`? Also, I don't see the need to test the `build` command line in addition to the PEP517 hook functions. Can you elaborate?

Reply:
Also, the Sage jobs in CI fail on this test, but there isn't enough in the logs to understand what goes wrong.

Reply:
The idea here is several projects could be tested. I wasn't very familiar with what would be good from the existing test suite, so I just picked one that looked important; the idea is more could be added. I can add a more if you want. If we never want more, the parametrization can be dropped.

The idea behind the virtual environment is to get an environment without the test requirements of the package (like ninja!) installed. However, due to the isolated build, I am wondering if it's really necessary, as long as you trust the isolated build to work. I might try it, maybe it wasn't needed?

Reply:
> build command line in addition to the PEP517 hook functions

Because build creates an isolated environment and downloads ninja, while the hook functions will always be using ninja from the test requirements. It's an integration test to make sure it builds without ninja present.

Reply:
Don't the PEP517 parametrized tests test just this?

Reply:
That's a (heavily mocked) unit test. This is an integration test to make sure you can actually build a package. The naming is rather terrible, I think this should be called "isolated_build" or something like that.

Reply:
Before this got added, test_wheel.py was much closer to an integration test. Now it's more noticeably not, since it will never add ninja, that's already a test dependency. This is a full integration test.

And to answer the question above, it isn't necessary to use virtualenv, it's still adding ninja to the isolated environment with subprocess.run.

I'd like to check the output and verify it's adding ninja. Not sure how to handle that test if there's already a ninja on the tester's system, though.

Reply:
And if you don't care about an integration test, I could drop it.

Reply:
Now I understand better. Thank you. However, note that running the test in a virtualenv in not enough to remove `ninja` from the environment if it is available in `$PATH` and on Linux I expect `ninja` to be available on most CI setups and on macOS too, probably. The check implemented here would thus find `ninja` and do not instruct `build` to install it. To remove `ninja` from the environment you need an higher degree of trickery, which I'm not sure is worth implementing.

Reply:
> And if you don't care about an integration test, I could drop it.

I think the integration test is valuable, better one test more than one test less, but I'm not sure the test tests does what you say.

Reply:
No, it wasn't to get ninja off the system environment (and still isn't), it was to get ninja out of the local environment. But since we are using an isolated build anyway (with build), that's just an extra layer that wasn't needed, and has been dropped - this is a subprocess.run now[^1]. Currently, it does pull in ninja if I uninstall it locally, but I'm assuming it's present in the CI so using that instead of ninja. (which is nice as well, but it would be better if we could control it).

Maybe a simple 1/2 solution would be to check to see if ninja is on the system, then require or disallow "ninja" in the build output based on that. Not ideal to adapt to the test system instead of of adapting the system to the test, but better than not caring at all if it adds ninja.

[^1]: In scikit-build-core, I also install it after building it, which does require the virtualenv.

Reply:
If it's important to remove it from the global system environment, maybe you can just create "poisoned" fake executables and prepend them to the system PATH, which report version 0.1 and/or exit with a failing return status?

Reply:
I see. If this is the goal, choosing `scipy-like` as a test project may not be the best choice, as it has extra build requirements that need to be installed in the isolated environment. Maybe a test project without extra dependencies could be used.

Anyhow, the test is failing in the CI. Can you adjust it so that the output of the executed command is spitted out in case of test failure so the situation can be debugged? Thank you!

Reply:
It does print out the result if it fails:

```
---------------------------- Captured stderr setup -----------------------------
/sage/local/var/lib/sage/venv-python3.10/bin/python3: No module named build
```

(Only visible in the logs artifact, since it's truncating the output for the GHA log)

Reply:
I don't know how to tell sage CI to add build to the venv it's creating, though. Doesn't seem to be using `.[test]` or nox.

Reply:
I looked at the GitHub Actions workflows involved and I didn't see anything that allows to specify python packages as test requisites. The last to touch the sage CI jobs definitions is @mkoeppe Maybe he knows more...

Reply:
The relevant file in Sage is build/pkgs/meson_python/dependencies_check [build/pkgs/meson_python/dependencies_check](https://github.com/sagemath/sagetrac-mirror/compare/develop...u/mkoeppe/numpy_1_23_x__scipy_1_9_x#diff-5a691b20a9521d9e7de0a407f41c8a473a55cf4549d3a2c66c4dbc0aacc59c81)

Reply:
but the easiest way to make `build` available is to add `python_build` before `meson_python` in https://github.com/mesonbuild/meson-python/blob/main/.github/workflows/ci-sage.yml#L110

Reply:
So I believe at least `pyproject-metadata` needs to be added to https://github.com/sagemath/sagetrac-mirror/blob/853a6995c01958d6ec26c8c68cb266944eb8dc5b/build/pkgs/meson_python/dependencies_check? Or I could just have these tests get skipped if the dependencies are not present, that would work too, though not as reassuring.

Reply:
@mkoeppe can you let us know what is the workflow to submit the required changes?

Reply:
I can also add a importorskip and "fix" the test by not testing on sage CI because it's missing dependencies. Then this could be added later. @mkoeppe please advise. If I don't hear anything, I'll probably go that route when I add `$NINJA`.